src/sentry/api/serializers/rest_framework/dashboard.py

```diff
@@ -0,0 +1,186 @@
+
+        existing_widgets = DashboardWidget.objects.filter(dashboard=instance, id__in=widget_ids)
+        existing_map = {widget.id: widget for widget in existing_widgets}
+
+        # Remove widgets that are not in the current request.
+        self.remove_missing_widgets(instance.id, widget_ids)
+
+        # Get new ordering start point.
+        next_order = get_next_dashboard_order(instance.id)
+
+        for i, data in enumerate(widget_data):
+            widget_id = data.get("id")
+            if widget_id and widget_id in existing_map:
+                # Update existing widget.
+                self.update_widget(existing_map[widget_id], data, next_order + i)
+            elif not widget_id:
+                # Create a new widget.
+                self.create_widget(instance, data, next_order + i)
+            else:
+                raise serializers.ValidationError(
+                    "You cannot use existing widgets that are not part of this dashboard."
+                )
+
+    def remove_missing_widgets(self, dashboard_id, keep_ids):
+        """
+        Removes current widgets belonging to dashboard not in keep_ids.
+        """
+        DashboardWidget.objects.filter(dashboard_id=dashboard_id).exclude(id__in=keep_ids).delete()
+
+    def create_widget(self, dashboard, widget_data, order):
+        widget = DashboardWidget.objects.create(
+            dashboard=dashboard,
+            display_type=widget_data["displayType"],
+            title=widget_data["title"],
+            order=order,
+        )
+        for i, query in enumerate(widget_data.pop("queries")):
+            DashboardWidgetQuery.objects.create(
+                widget=widget,
+                fields=query["fields"],
+                conditions=query["conditions"],
+                name=query.get("name", ""),
+                interval=query.get("interval", "5m"),
+                order=i,
+            )
+
+    def update_widget(self, widget, data, order):
+        widget.title = data.get("title", widget.title)
+        widget.display_type = data.get("displayType", widget.display_type)
+        widget.order = order
+        widget.save()
+
+        if "queries" in data:
+            self.update_widget_queries(widget, data["queries"])
+
+    def update_widget_queries(self, widget, data):
+        query_ids = [query["id"] for query in data if "id" in query]
+        self.remove_missing_queries(widget.id, query_ids)
+
+        existing = DashboardWidgetQuery.objects.filter(widget=widget, id__in=query_ids)
+        existing_map = {query.id: query for query in existing}
+
+        for i, query_data in enumerate(data):
+            query_id = query_data.get("id")
+            if query_id and query_id in existing_map:
+                self.update_widget_query(existing_map[query_id], query_data, i)
+            if not query_id:
+                self.create_widget_query(widget, query_data, i)
```

Comment:
This also makes `N` rows, can the queries be split up based on if it needs to be updated/created? That way, the new queries can be bulk created.